### PR TITLE
Capture form inputs

### PR DESCRIPTION
**Pull request by Luciam91**
Left to do:

- [x] Filter for Credit Card details
- [x] Fix an issue where it fires _twice_
- [x] Actually do the form submit it _should do_
- [x] Apply to forms if new ones appear

### CODE REVIEW

File: src/hooks/useFormCollector.ts [hunk 1]
```diff
@@ -1,40 +1,108 @@
...
+
+const bannedTypes = ['password', 'submit']
+const bannedFieldPartialNames = [
+  'expir',
+  'cvv',
+  'cvc',
+  'csv',
+  'pin',
+  'pass',
+  'card'
```
Comment (Luciam91) [inline src/hooks/useFormCollector.ts:18]: I worry _slightly_ about this, but probably not enough to block merging

File: dist/client/types.d.ts [hunk 2]
```diff
@@ -9,12 +9,6 @@ export declare type CollectorUpdate = {
     referrer?: Referrer | undefined;
     elements?: PageElement[] | undefined;
     account?: Account | undefined;
-    form?: Form | undefined;
-};
-export declare type Form = {
```
Comment (Luciam91) [inline dist/client/types.d.ts]: Isn't this needed so that the types are correct?

File: src/hooks/useFormCollector.ts (hunk 1 above)
Comment (AdenFraser) [reply src/hooks/useFormCollector.ts:18]: MAB uses `csc` on their flows from what I can see, you'll need to capture that

File: src/hooks/useFormCollector.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+export default function useFormCollector() {
+  const { mutateAsync: collect } = useCollectorMutation()
+  const { visitor } = useVisitor()
+  const [haveFormsBeenDetected, setHaveFormsBeenDetected] =
+    useState<boolean>(false)
+
+  useEffect(() => {
+    if (isUndefined('document')) return
+
+    const intId = setInterval(() => {
```
Comment (AdenFraser) [inline src/hooks/useFormCollector.ts:47]: If I'm reading this correctly, will this:

- Check whether forms exist
- If they don't, check again every second until they do
- If they do, stop checking?

If so, does that put us at risk that a booking form/widget that loads in more than 1000ms on a page where other form inputs exist (like a search) are never picked up?

Comment (EdPutans) [reply src/hooks/useFormCollector.ts:47]: Very valid point.

Comment (EdPutans) [reply src/hooks/useFormCollector.ts:50]: sorted :)

File: src/hooks/useFormCollector.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    3.2 filter out the fields based with sensitive information
+    3.3 collect the data and send it to the server
+    3.4 without waiting for the server response, optimistically submit the form after a 
+    delay of `submitionDelay` ms
+* 4. 
+ */
+export default function useFormCollector() {
+  const { mutateAsync: collect } = useCollectorMutation()
+  const { visitor } = useVisitor()
+  const [nbDetectedForms, setNbDetectedForms] = useState<number>(0)
```
Comment (AdenFraser) [inline src/hooks/useFormCollector.ts:45]: What does 'nb' stand for? Is it an acronym?

**Comment by AdenFraser** [comment]
This approach is fine. A better way would be to diff the NodeList from your document query selector, as this still has the chance of side effects, but the chances are very slim.

File: src/hooks/useFormCollector.ts (hunk 4 above)
Comment (EdPutans) [reply src/hooks/useFormCollector.ts:45]: number :D

**Review by EdPutans — APPROVED** [review]

**Pull request merged**